app/scripts/models/flow.js

```diff
@@ -0,0 +1,130 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * A model to represent the somewhat nebulous concept of an
+ * "event flow". A single flow could be a user connecting a
+ * new device to their Sync account, for instance.
+ *
+ * Each flow has two properties.
+ *
+ *   * flowId: 32-byte identifier hex-encoded as a string.
+ *   * flowBeginTime: server-generated timestamp.
+ *
+ * When a new flow is begun, it should result in an activity
+ * event being logged by the server. To that end, we push a
+ * a `flow.begin` onto the metrics event array.
+ *
+ * There is some mess surrounding when to begin a new flow,
+ * which necessitates equivalent mess in this model. It runs
+ * something like this:
+ *
+ *   1. A user arriving at /signin or /signup constitutes the
+ *      beginning of a flow, so a corresponding `flow.begin`
+ *      event must be logged.
+ *
+ *   2. A user refreshing the page on /signin or /signup does
+ *      not constitute a new flow, so events must not be logged
+ *      in those cases. This is achieved by saving the flowId
+ *      in local storage.
```

Comment:
Although it seems weird on the face of it, when you consider how the metrics are calculated I don't think it is.

If you look at the last query here:

https://github.com/mozilla/fxa/blob/master/features/FxA-70-kpi-dashboards/README.md#proposed-queries

...the percentage of completed device connections will be affected, because these cases would generate extra, unfinished `flow.begin` events. So my gut is telling me that even though it means the time-to-device-connection metric would be skewed by the current system for these cases, that is the right way to go.

But I'll defer to @rfk, who is the expert on this stuff. :grin: